File: rust/2019/2-1.rs
use std::io::{stdin, Read};

type MemoryRef<'a> = &'a mut Vec<usize>;
type PositionRef<'a> = &'a mut usize;

fn parse_input(input: &String) -> Vec<usize> {
    input
        .split(',')
        .map(|x| x.parse().expect("Failed to parse input"))
        .collect()
}

fn step(position: PositionRef) {
    *position += 4;
}

fn run_operation<F>(memory: MemoryRef, position: PositionRef, func: F)
where
    F: Fn(usize, usize) -> usize,
{
    let x = memory[memory[*position + 1]];
    let y = memory[memory[*position + 2]];
    let out = memory[*position + 3];
    memory[out] = func(x, y);
}

fn run(memory: MemoryRef, position: PositionRef) -> usize {
    let operation = memory[*position];
    match operation {
        1 => {
            // Add
            run_operation(memory, position, |x, y| x + y);
            step(position);
            run(memory, position)
        }
        2 => {
            // Multiply
            run_operation(memory, position, |x, y| x * y);
            step(position);
            run(memory, position)
        }
        99 => {
            // Halt
            memory[0]
        }
        _ => panic!("Invalid operation {} at position {}", operation, position),
    }
}

fn solve(input: &String) -> usize {
    let mut memory = parse_input(&input);
    let mut position = 0;

    memory[1] = 12;
    memory[2] = 2;

    run(&mut memory, &mut position)
}

fn main() {
    let mut input = String::new();

    stdin()
        .read_to_string(&mut input)
        .expect("Failed to read stdin");

    let solution = solve(&input);

    println!("{}", solution);
}

File: haskell/2019/2-1.hs
import Control.Monad.ST (ST, runST)
import Data.STRef (STRef, modifySTRef, newSTRef, readSTRef)

type MemoryRef s = STRef s [Int]

type PositionRef s = STRef s Int

type Context s = (MemoryRef s, PositionRef s)

modifyAt :: Int -> (a -> a) -> [a] -> [a]
modifyAt i f xs = take i xs <> [f (xs !! i)] <> drop (i + 1) xs

parseInput :: String -> [Int]
parseInput input = read ("[" <> input <> "]")

get :: MemoryRef s -> Int -> ST s Int
get ref index = (!! index) <$> readSTRef ref

put :: MemoryRef s -> Int -> Int -> ST s ()
put ref index value = modifySTRef ref (modifyAt index (const value))

step :: PositionRef s -> ST s ()
step positionRef = modifySTRef positionRef (+ 4)

runOperation :: MemoryRef s -> Int -> (Int -> Int -> Int) -> ST s ()
runOperation memoryRef position fn = do
  x <- get memoryRef =<< get memoryRef (position + 1)
  y <- get memoryRef =<< get memoryRef (position + 2)
  out <- get memoryRef (position + 3)
  put memoryRef out (fn x y)

run :: Context s -> ST s Int
run context@(memoryRef, positionRef) = do
  position <- readSTRef positionRef
  operation <- get memoryRef position
  case operation of
    1 -> do
      -- Add
      runOperation memoryRef position (+)
      step positionRef
      run context
    2 -> do
      -- Multiply
      runOperation memoryRef position (*)
      step positionRef
      run context
    99 ->
      -- Halt
      get memoryRef 0
    invalid ->
      error ("Invalid operation " <> show invalid <> " at position " <> show position)

solve :: [Int] -> Int
solve xs = runST $ do
  memoryRef <- newSTRef xs
  positionRef <- newSTRef 0
  let context = (memoryRef, positionRef)
  put memoryRef 1 12
  put memoryRef 2 2
  run context

main :: IO ()
main = print . solve . parseInput =<< getContents
